wagtail/contrib/settings/views.py

```diff
@@ -82,118 +86,280 @@ def redirect_to_relevant_instance(request, app_name, model_name):
+        app_name = self.kwargs["app_name"]
+        model_name = self.kwargs["model_name"]
+        pk = self.kwargs["pk"]
 
-    if not user_can_edit_setting_type(request.user, model):
-        raise PermissionDenied
+        self.model = get_model_from_url_params(app_name, model_name)
 
-    setting_type_name = model._meta.verbose_name
-    edit_handler = get_setting_edit_handler(model)
-    form_class = edit_handler.get_form_class()
-    site: Optional[Site] = None
-    site_switcher = None
+        if issubclass(self.model, BaseSiteSetting):
+            return self.model.for_site(get_object_or_404(Site, pk=pk))
+        else:
+            return get_object_or_404(self.model, pk=pk)
 
-    if issubclass(model, BaseSiteSetting):
-        site = get_object_or_404(Site, pk=pk)
-        form_id = site.pk
-        instance = model.for_site(site)
+    def get_form_class(self):
+        return get_setting_edit_handler(self.model).get_form_class()
 
-        if request.method == "POST":
-            form = form_class(
-                request.POST, request.FILES, instance=instance, for_user=request.user
-            )
+    def get_edit_url(self):
+        return "wagtailsettings:edit"
 
-            if form.is_valid():
-                with transaction.atomic():
-                    form.save()
-                    log(instance, "wagtail.edit")
-
-                messages.success(
-                    request,
-                    _("%(setting_type)s updated.")
-                    % {
-                        "setting_type": capfirst(setting_type_name),
-                        "instance": instance,
-                    },
-                )
-                return redirect("wagtailsettings:edit", app_name, model_name, site.pk)
-            else:
-                messages.validation_error(
-                    request, _("The setting could not be saved due to errors."), form
-                )
-        else:
-            form = form_class(instance=instance, for_user=request.user)
-
-        edit_handler = edit_handler.get_bound_panel(
-            instance=instance, request=request, form=form
-        )
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
 
-        media = form.media + edit_handler.media
+        site = None
+        site_switcher = None
 
-        # Show a site switcher form if there are multiple sites
-        if Site.objects.count() > 1:
-            site_switcher = SiteSwitchForm(site, model)
-            media += site_switcher.media
+        if issubclass(self.model, BaseSiteSetting):
+            site = get_object_or_404(Site, pk=self.kwargs["pk"])
+            instance = self.model.for_site(site)
 
-    elif issubclass(model, BaseGenericSetting):
-        queryset = model.base_queryset()
+            # Show a site switcher form if there are multiple sites
+            if Site.objects.count() > 1:
+                site_switcher = SiteSwitchForm(site, self.model)
```

Comment:
The media for the `SiteSwitchForm` in the original code seems to be missing here.

```suggestion
                site_switcher = SiteSwitchForm(site, self.model)
                media = context.get("media") + site_switcher.media
```

(and don't forget to update `"media"` in `context.update()`)